Design portable tail calls with helpers

janvorli opened the pull request:
Currently we don't handle non-trivial cases of tail calls on Unix due to the non-portable and
complex way it is implemented. This pull request describes a design of a portable way of
handling these cases so that it can be used uniformly over all our jitted and AOT (ahead of time)
platforms.

janvorli commented:
This change is related to #4971, #2556, https://github.com/dotnet/corert/issues/1683

janvorli commented:
CC: @jkotas, @BruceForstall

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,260 @@
+* The stack should be unwindable at any spot during the tail calls to properly support sampling profilers and similar tools.
+* Stack walk during GC must be able to always correctly report GC references. 
+* It should work in all cases except those where a tail call is not allowed as described in the ECMA 335 standard section III.2.4
+## Requirements
```
The formatting for this does not look right in the rendered md view.

jkotas replied:
(Multiple places)

janvorli commented:
I've fixed the formatting issue and amended the commit. VS Code preview was not showing the problem.

BruceForstall commented:
cc @sivarv

BruceForstall commented:
LGTM

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+* At the time of the tail call, the caller generates a vararg list of all arguments of the tail called function and then calls JIT_TailCall runtime function. It passes it the copying routine address, the target address and the vararg list of the arguments.
+* The JIT_TailCall then performs the following:
+  * It calls RtlVirtualUnwind twice to get the context of the caller of the caller of the tail call to simulate the effect of running epilog of the caller of the tail call and also its return.
+  * It prepares stack space for the callee stack arguments, a helper explicit TailCallFrame and also a CONTEXT structure where the argument registers of the callee, stack pointer and the target function address are set. In case the tail call caller has enough space for the callee arguments and the TailCallFrame in its stack frame, that space is used directly to for the callee arguments. Otherwise ...
```
>to for  [](start = 361, length = 7)

Nit: "direct for" 
"to" is not required.

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+The new way of handling tail calls using helpers was designed with the following objectives:
+* It should be cheap to port to new platforms, architectures and code generators
+* It needs to work in both jitted and AOT compiled scenarios
+* It should support platforms where runtime code generation is not possible
```
>It should support platforms where runtime code generation is not possible [](start = 2, length = 73)

Is this object different from the previous one that mentions " and AOT compiled scenarios"?

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+  * Get the arguments buffer from the thread local storage, extract the arguments, release the buffer and call the target function. The frame of the CallTarget thunk ensures that the arguments of the target are GC protected until the target function returns or tail calls to another function.
+  * Pop the TailCallHelperStack entry from the TailCallHelperStack of the current thread.
+  * Check the ChainCall flag in the TailCallHelperStack entry. If it is set, run epilog and jump to the TailCallHelper.
+  * If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function.
```
>If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function. [](start = 4, length = 151)

How is the return value GC-protected?
Say the callee returns a ref/byref or on amd64 unix returns a struct containing two ref/byref fields in return register(s) RAX/RDX.  The control of execution returns to CallTarget Thunk (perhaps going through a series of CallTargetThunk->TailCallHelper calls if the tail calls are chained).  Can GC kick-in while executing inside TailCallHelper or CallTarget Thunk?

sivarv replied:
Please include answers to the above question as well in this document.

---
In reply to: [92237287](https://github.com/dotnet/coreclr/pull/8596#discussion_r92237287) [](ancestors = 92237287)

janvorli replied:
I believe there is no magic and regular machanism of GC protection of the return value would work here. When the CallTarget thunk is returning, it is the same case as if the original caller of the tail call chain called the CallTargetThunk which in turn called the target function. The CallTargetThunk is a regular managed code method. The return value from the last function of the tail call chain always travels that way, there is no TailCallHelper in between in that case.
Does it make sense?

sivarv replied:
For regular mechanism of GC protection to work  CallTargetThunk() should have the same return type as the return type of the method that is tail called.   That is the point I have missed.
You might want to explicitly mention this under "Helper functions" section against CallTargetThunk, though it is implicit from the above line.

---
In reply to: [92259054](https://github.com/dotnet/coreclr/pull/8596#discussion_r92259054) [](ancestors = 92259054)

sivarv replied:
On further thinking, it looks it will work for single-level tail calls but not for chained tail calls.

Let us consider an example tail call chain

A() -> calls B() --> tail call C() 

Assume further B(), C() return a gc-type return value.

Here is how the call stack looks while executing inside C().

A() -> CallTarget of C --> C()

Say C() in turn tail call into D(), which also return a gc-type return value.

C() would call StoreArgument thunk to store args for D in TLS and jumps to TailCallHelper after executing its epilog.  Essentially call stack looks as

A() --> CallTarget of C --> TailCallHelper()

TailCallHelper() finds that previous stack frame is CallTarget() and hence marks it as chained and returns to CallTarget of C().  This will appear to CallTarget as if C() has returned, but return register do not contain a valid gc-ref.  
Say if GC kicks in immediate after returning to CallTarget of C.  As per gc-info of CallTarget of C, return reg (say RAX) contains a valid gc-ref but it does not. 

Q1: Won't it lead to a crash when GC tries to use in invalid gc-ref in rax and mark it? Is my understanding right here?

Call stack after TailCalHelper() returns is

A() --> CallTarget of C

CallTarget on finding ChainCall Flag, executes its epilog and jumps to TailCallHelper.  What is the return value of TailCallHelper()?  Since it is common for all tail call targets, I am assuming to be void.  

Q2: Can GC kick-in while executing inside TailCallHelper()?

Call stack at this point is

A() --> TailCallHelper()

TailCallHelper() in turn now jumps to CallTarget of D().  The call stack now is

A() --> CallTarget of D --> D()



---
In reply to: [92481487](https://github.com/dotnet/coreclr/pull/8596#discussion_r92481487) [](ancestors = 92481487,92259054)

janvorli replied:
I just got another idea. What if we marked the CallTarget thunk as non-interruptible function? It seems it would solve the problem with the return from TailCallHelper not having a valid return value.

jkotas replied:
> For the hidden return buffer argument`

The hidden return buffer argument works and is reported to the GC in the same way as normal byref argument (e.g. what you get if you write `out LargeStruct val` in C#). I think the hidden argument should be passed to StoreArguments, StoreArguments should store it as any other byref argument, and then CallTarget just needs to store result into it once the actual call returns. It keeps the TailCallHelper  helper simple.

> non-interruptible function

The non-interruptible regions (or conservative reporting) may be potentially interesting for getting rid of the GC descriptors, but they make the contract with the JIT more complex. I do not think it is worth going there.

jkotas replied:
> the GC kicking in while executing inside TailCallHelper - it can happen

It cannot happen. TailCallHelper is unmanaged code running in cooperative mode. If the GC wants to run, it needs to wait for the execution to leave the TailCallHelper.

There is related problem: We should make sure that there is enough safe points for GC to run. The fast tailcalls achieve it today by marking all methods with tailcalls as fully interruptible. It has GCInfo encoding size hit, but it works fine otherwise - we can do the same here.

janvorli replied:
Ah, right, I've forgotten we are in cooperative mode in this helper.
As for the safe points, isn't the fact that we are returning to the previous CallTarget thunk at each tail call in a chain enough to ensure enough safe points?

jkotas replied:
>  we are returning to the previous CallTarget thunk at each tail call 

Yes, you are right - it should be good enough as a safe point.

jkotas replied:
I do not think that the idea that I have proposed is the best solution after mulling about it some more. I would vote for TailCallHelper having the same return type as the target of the tail call. @sivarv @BruceForstall What do you think?

sivarv replied:
I am still trying to understand how struct return via retBuf arg are handled in the new tail call mechanism.  I am writing down my understanding below if we were to implement second idea from @jkotas.

say A() -> call B() -> tail calls C()
where B() and C() return a value type passed via hidden RetBuf arg.

A() would allocate a struct local for return value and pass a byref to it to B().
B() would first call StoreArgument() IL thunk with all the arguments as if it were calling C() except that it considers its return value to be void.  

If B() were to call C() normally, jitted code of B() is free to allocate another local to receive return value from C and copy that back to RetBuf that A() has passed to B().  In case of B() tail calling C(), we want B() to pass Retbuf arg that A() passed to it to StoreArgument() thunk.  Note C() might require other hidden args, example generic context those too needs to be passed to StoreArgument() thunk.  Vararg cookie may not be interesting since we don't support tail calling vararg methods.  There could be other non-standard hidden args (e.g. Virtual Stub Dispatch (VSD) param R11 on Amd64).

Side Question: How do we see that CallTarget IL thunk passing other hidden args like generics cookie to C(), given CallTarget IL thunk is JIT compiled like any other managed method?  Also , we need layout how the new proposal works tail calling VSD calls that require hidden arg passed in R11.

StoreArgument() thunk stores args in TLS and branches to TailCallHelper, which in turn copies args from TLS to stack and branches to CallTarget() IL thunk.  Given CallTarget() is an IL thunk jit compiled like any other managed method, while calling C(), it would allocate a local for receiving retVal from C() and copy that to the RetBuf passed by B() in args.

A() -> CallTarget() of C -> C()

In case of single level tail call C() would return struct retval by copying into RetBuf that was passed by A() to B() and it would work correctly.  

Say C() in turn tail calls into D.
C() --> StoreArgument() thunk - C() would have to pass RetBuf arg that in turn received from its caller to D()
C() --> branches to TailCallHelper
TailCallHelper --> marks as chained tail call and returns to CallTarget() of C by executing its epilog.
Before returning TailCallHelper is required to zero init at least gc-ref fields of RetBuf arg
CallTarget() of C --> copy return value to RetBuf that it received from B() and branches to TailCallHelper after executing its epilog.
TailCallHelper --> branches to CallTarget of D
CallTarget of D --> calls D() and copies the return value it received from D() into RetBuf arg that it received from C() which is same as RetBuf arg C() received from B() which is same as RetBuf arg it received from A().
Finally when CallTarget of D returns to A(), A() would find the return value in RetBuf arg that it passed and things work correctly.

Regarding @janvorli's idea of having TailCallHelper as having the same return type as the target of the tail call:
I thought TailCallHelper is an asm helper that is common to all tail call targets. This would mean now we will have one TailCallHelper per tail call target, right?  

Revisiting above example of A() -> call B() -> tail calls C().
B() -> StoreArgument() thunk - so B() would not pass RetBuf arg in arguments to the thunk (is that correct?)
B() -> branches to TailCallHelper by passing it RetBuf arg that it received from A()
TailCallHelper() -> setup args for CallTarget of C() on stack and branches to it
CallTarget of C() -> calls C() passing all required args

Question: Please help me understand what advantages we see having TailCallHelper having the same return type as target of the tail call over passing RetBuf to StoreArguments Thunk?

janvorli replied:
Now regarding the steps you have described for the solution when we consider the TailCallHelper as having the same return value as the tail call target:
>B() -> StoreArgument() thunk - so B() would not pass RetBuf arg in arguments to the thunk (is that correct?)

Correct
>B() -> branches to TailCallHelper by passing it RetBuf arg that it received from A()

Correct
>TailCallHelper() -> setup args for CallTarget of C() on stack and branches to it

Incorrect. The TailCallHelper would pass just the hidden return buffer argument to the CallTarget of C. The CallTarget of C has no regular parameters.
> CallTarget of C() -> calls C() passing all required args

Incorrect. The CallTarget of C extracts the regular parameters from the TLS buffer and then calls C

janvorli replied:
As for the other hidden arguments:
I believe the jitter would generate them automatically. Since the CallTarget is a regular managed method, it would call the target the same way as the caller of the tail call would call it if it was not a tail call, but a regular call. So I assume the jitter would prepare all of these extra parameters when compiling the call to the target function in the CallTarget.
I have just found though that there is one thing that I have forgotten to describe. When the target of the tail call is an instance method, we also need to store the "this" pointer in the arguments buffer so that we have it available when calling the target method.

sivarv replied:
_I believe the jitter would generate them automatically. Since the CallTarget is a regular managed method, it would call the target the same way as the caller of the tail call would call it if it was not a tail call, but a regular call._

Yes, for that reason, I am inclined to have all args passed to StoreArgument IL thunk (including RetBuf, thisptr, generic context, VSD parm), unless there are compelling reason(s)/advantages because of which we want to have some arguments special cased and passed to TailCallHelper.

jkotas replied:
> I am inclined to have all args passed to StoreArgument IL thunk

The problem is what to do with these hidden arguments in the CallTarget method part. There are no natural constructs in IL to fetch them, and so we would need to invent a new intrinsics or something to work with them.

janvorli replied:
@sivarv I am not sure I understand your response. I have said that I believe there is no need to pass those hidden arguments to StoreArguments (except of this pointer) or to do anything explicit about them. It would just work - the call to the target of the tail call would get them automatically. I don't see why we would need to explicitly handle the generic context or VSD parm here. The CallTarget is a function like any other, so when it calls the target function, it calls it the same way as if that function was at any other place in the code. So the jitter has to take care of providing those arguments no matter where the function is called from.

sivarv replied:
I was implying the same question earlier. pasting below

_Side Question: How do we see that CallTarget IL thunk passing other hidden args like generics cookie to C(), given CallTarget IL thunk is JIT compiled like any other managed method? Also , we need layout how the new proposal works tail calling VSD calls that require hidden arg passed in R11._

May be we should also have details on how CallTarget IL thunk looks like for calling a variety methods.

sivarv replied:
I had a quick offline chat with @jkotas.

E.g. A() -> calls B() -> tail calls C()

VSD stub param - I see that this param will be materialized by JIT when CallTarget IL thunk calls the actual callee method (C() in the above example).
Generic context - this can only be provided by the original caller in some cases.  This too would need special handling like RetBuf or somehow we have to make CallTarget to have the same generic-ness as caller method (B() in the above example)

janvorli replied:
@sivarv can you please show me an example of the case where we would need to pass the generic context?

jkotas replied:
```
static void foo<T>()
{
    ...
    tailcall.    bar<List<T>>(1,2,3,4,5,6,7,8);
}

static void bar<T>(int a1, int a2, int a3, int a4, int a5, int a6, int a7, int a8)
{
   ....
}
```

janvorli replied:
Hmm, ok I think I can see it. We would need to do that in all cases where the callee generic parameters are not known at the compile time, right?

janvorli replied:
Ah, thanks, I can see it now. 
So I think that we can do it this way. We would consider TailCallHelper to have the same generic-ness as the target of the tail call and the CallTarget would have the same generic-ness too.

janvorli replied:
@sivarv here are examples of the il stubs for several cases (member method, static method, generic method, method of generic class, interface):
https://gist.github.com/janvorli/b9699c05f1457957bda8494eb3884034
I have tested they all work correctly (for the testing purposes there was no GC protection of references while they are in the buffer)

sivarv replied:
@janvorli - We intend to make CallTarget IL thunk and TailCallHelper to have the same generic-ness as the target of the tail call.  For my benefit can you please elaborate what it would mean for VM and JIT?

Since CallTarget thunk is IL generated, while generating I guess VM could make it a generic method with the same number of generic params as that of target of the call?  TailCallHelper is an asm helper.  What do we mean by consider TailCallHelper as having the same generic-nesss?  Does JIT have to special case call to TailCallHelper and pass it generic context param?

sivarv replied:
Tail call profiler callback:
where should tail call profiler hook be generated in the new mechanism?  After the call to StoreArgument thunk call but before branching to TailCallHelper?

Are there any other profiler ABI requirements (e.g. tail call args should be in certain place etc) before the tailcall callback is made?  I guess not.

jkotas replied:
The JIT should decide based on the ABI:
- Desktop: Always the existing one
- CoreCLR: Switch all platforms to the new one over time (maybe except x86 - can stay on the existing plan)
- CoreRT: Always the new one

jkotas replied:
The interesting question is what value is going to be passed as the hidden argument to TailCallHelper if there is one, and how does the JIT manufacture it. It can be either the generic context for CallTarget, or the generic context for actual target.

janvorli replied:
@jkotas I assume you mean x86 Windows, right? I don't see why we would want to do it for x86 Linux. I think the existing varags mechanism would have the same problem as on other architectures.

sivarv replied:
I think it is doable.   If the target of tail call has a retBuf arg, then the same should be passed to TailCallHelper.  Similarly if the target of tail call has a GenericContext Param, the same should be be passed to TailCallHelper.
TailCallHelper would in turn pass these parameters to CallTarget IL thunk, which would in turn pass will pass it to the target method.

Would that mean that we will have the following TailCallhelpers?
TailCallHelper
TailCallHelper_RetBuf
TailCallHelper_GenericCtx
TailCallHelper_RetBuf_And_GenericCtx

jkotas replied:
> I assume you mean x86 Windows, right? I don't see why we would want to do it for x86 Linux.

There are no varargs involved on x86. It is just a single assembly helper, pretty similar to your TailCallHelper. The difference is that the assembly helper takes care of storing and restoring the arguments. It is possible on x86 because of the x86 calling convention is very simple. The assembly helper takes an extra int that describes the size of the argument block pushed on the stack, and moves that block to the right spot.

jkotas replied:
> TailCallHelper would in turn pass these parameters to CallTarget IL thunk, which would in turn pass will pass it to the target method.

The generic context is potentially changing per generic method. The generic context for `CallTarget<T>` would be different from the generic context of `ActualTarget<T>` if you are staying on regular IL plan.

janvorli replied:
@sivarv I think that having these four flavor of the helper instead of one would be reasonable. Altough it could work with a single helper too given the fact that two arguments can be passed in registers on all the platforms and so I can just pass the first two argument register passed to the TailCallHelper untouched to the CallTarget, not even caring about what's in them. This might not be true on some AOT targets though, so I would put both of these options to the doc.

janvorli replied:
@sivarv as for your questions:
> Will there be any issue because both existing and new tail call mechanism needs to be supported by VM?

I don't see any problems here. The VM is built for the target platform, so it would be built with support for one of the mechanisms, based on the target platform.
> How does JIT will get to know which tail call helper mechanism to use?

If I recall it correctly, there is already a mechanism that allows the JIT to figure out if it is generating code for CoreRT or CoreCLR. Maybe we can extend that mechanism or provide a way to query it via the JIT to EE interface.

janvorli replied:
Regarding
> if the target of tail call has a GenericContext Param, the same should be be passed to TailCallHelper.
TailCallHelper would in turn pass these parameters to CallTarget IL thunk, which would in turn pass will pass it to the target method.

Since @jkotas has pointed out that the generic context of the CallTarget thunk can be different from the generic context of the actual tailcall target, the generic context passed to the TailCallHelper should be the context of the CallTarget thunk. The TailCallHelper would pass the generic context to the CallTarget and the jitter would then create the right generic context for the actual call to the target. So there should be no special handling of the CallTarget thunk by the jitter.

sivarv replied:
I am not sure that I understand how this works from JIT point of view.  

e.g.
A() -> call B() --> tail call C()

Say we are jit compiling B().  At this time JIT knows the generic context that needs to be passed to C().  How will JIT know the generic context that needs to be passed to CallTarget IL thunk?

It would be good to construct an example case where the generic context that is to be passed to CallTarget thunk is different from the one to be passed to target of tail call.

janvorli replied:
Regarding the tail call profiler callback:
> where should tail call profiler hook be generated in the new mechanism? After the call to StoreArgument thunk call but before branching to TailCallHelper?

It sounds like a reasonable place. I don't think the tail call hook relies on anything that would depend on whether it is called before or after storing the arguments.

janvorli replied:
@sivarv one of the cases where the generic context cannot be the same for CallTarget and the tailcall target is the case of a generic method in a non-generic class. In that case, the generic context is the MethodDesc of the method and you cannot have two different methods having the same MethodDesc. 
So we really need to pass the generic context of the CallTarget and not the one of the tailcall target to the TailCallHelper. 
As for your question:
> Say we are jit compiling B(). At this time JIT knows the generic context that needs to be passed to C(). How will JIT know the generic context that needs to be passed to CallTarget IL thunk?

I don't know how the generic context is figured out in the JIT, but since it can figure it out for a call to C, couldn't it use the same way to figure it out for the CallTarget?
Actually, what if we did things the way that the jitter would generate code as if it was jumping to the CallTarget (but jump to the TailCallHelper instead)? It seems that that would ensure that it would pass in the proper generic context of the CallTarget in a natural way.

sivarv replied:
VM indicates to JIT whether a call site requires an extra genericCtx param to be passed and also provides the context handle that is to be passed as generic context.

For the example `A() -> B()`

Say A() is getting JIT compiled and JIT is importing call to B().  JIT asks VM about `CORINFO_CALL_INFO` using `getCallInfo()` JIT-EE interface call.  CallInfo struct has `CORINFO_SIG_INFO` that provides details on signature of the call.

If  `sig->callConv` has CORINFO_CALLCONV_PARAMTYPE flag set, then such a call requires to be passed a genericCtx.  
Further callInfo struct two fields: contextHandle and whether exactContextNeedsRuntimeLookup. contextHandle encodes both the handle and a flag (CORINFO_CONTEXTFLAGS_METHOD or CORINFO_CONTEXTFLAGS_CLASS) to indicate whether the context handle is a method or class handle.  JIT derives genericCtx from contextHandle by clearing the flags.

We cannot use the same JIT-EE interface call - getCallInfo() -  for determining genericCtx that is to be passed to CallTarget IL thunk, because that interface call requires CORINFO_RESOLVED_TOKEN corresponding to the call.  While jit compiling A(), JIT has CORINFO_RESOLVED_TOKEN for call to B() and not to CallTarget IL thunk.

janvorli replied:
@sivarv thanks a lot for the details. But why cannot we get the CORINFO_RESOLVED_TOKEN for the CallTarget IL thunk the same way we would get it for the target of the tail call? Cannot we call the resolveToken to get it at the point where we are generating the code for the jump to the TailCallHelper?

sivarv replied:
CORINFO_RESOLVED_TOKEN struct is returned by VM while resolving an IL token using resolveToken() JIT-EE call.  In the above example while JIT compiling A(), in the IL stream JIT finds call token to B() and calls VM to resolves it.   To resolve a token JIT needs to supply: tokenContext, tokenScope, token and tokenType.  

CallTarget() IL thunk is an artificially synthesized method generated at runtime by VM.  How will JIT get its token and also required info for its token resolution?

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+### Helper data structures
+The tail calls use the following data structures:
+* Thread local storage for arguments. It stores the arguments of a tail call for a short period of time between the StoreArguments and CallTarget calls.
+* Arguments GC descriptor - descriptor of locations and types of managed references in the arguments.
```
> Arguments GC descriptor - descriptor of locations and types of managed references in the arguments. [](start = 1, length = 100)

On amd64 Unix, structs are passed directly in registers or on stack. Say the callee method tail called has struct args and those structs has GC ref/byref type fields (e.g. RuntimeTypeHandle).  Would GC descriptor contains details of gc ref/byref fields of the struct or  the GC-layout of value type needs to be constructed each time using its type handle during gc-reporting?  I am assuming the former.  Can you give some examples of what GC descriptor looks like while passing structs containing obj ref/byref fields?

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+* Return value and call target arguments are all either primitive types, reference types, or valuetypes with a single primitive type or reference type fields
+* The aligned size of call target arguments is less or equal to aligned size of caller arguments
+
+## Tail calls using a helper
```
Please also provide details on how gc-reporting of args and return value of callee is done in this mechanism for completeness.

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+  * Pointer to the next entry on the stack.
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
```
>The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage [](start = 2, length = 187)

On amd64 Windows, structs are passed by by-ref to a local on caller's stack frame.  This would mean while copying args, to TLS,  entire struct needs to be copied to TLS area.  This means GC descriptor needs to describe layout of struct args as well?
How are byref type params handled in general in the new mechanism?

sivarv replied:
Also it would be good to capture how this handled in the existing tail call helper mechanism in this doc.

---
In reply to: [92244721](https://github.com/dotnet/coreclr/pull/8596#discussion_r92244721) [](ancestors = 92244721)

janvorli replied:
The GC descriptor needs to describe all arguments that are references or byrefs and also all members of struct arguments that are references.
As for describing how that works in the existing tail call helper mechanism, I'll add something to this doc.

sivarv commented:
CC @dotnet/jit-contrib

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,266 @@
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
+* The caller executes its epilog, restoring stack pointer and callee saved registers to their values when the caller was entered.
```
Managed debugger:
My understanding is that under VS (or managed debugger), managed callstack hides call stacks of tail call helper/thunks and provides a logical call stack view

e.g. A() ->calls B() --> tail calls C
A() -> CallTarget() -> C()

At this point would managed call stack under VS be 
A() -> C()  
or one will also get to see CallTarget() in stack trace?

You might want to check what the existing tail call helper mechanism on Windows does in this case.

janvorli commented:
@sivarv, @jkotas I have updated the document based on the feedback from both of you and also added an example of IL code generated for the StoreArguments and CallTarget helpers.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,375 @@
+    ldloc.0
+    ldind.i // int x
+    ldloc.0
+    sizeof int32
```
If you just add sizeof(int32) here, you will hit a potentially misaligned access fetching the next argument.

janvorli replied:
Right, I'll fix it.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,375 @@
+    stloc.0
+
+    ldloc.0
+    ldobj object // generic context
```
The generic context is not object. It is IntPtr.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,375 @@
+    ldloc.0
+    ldobj object // T t
+
+    calli Point (object, object, int32, object)
```
Should this be regular call instead? (I do not see where the address for calli is coming from.)

jakobbotsch replied:
I know this is years old at this point, but it seems hard (/potentially fragile) to make this `calli`. For example, on x86 the type context comes after user args, so this would need to be special cased there in the JIT and the AOT scenario.

I am not particularly sure I understand why this cannot just be a direct `call` (which then may also be potentially inlined). There is a large discussion about it above, but I am not sure I understand why this should not be technically possible. After all the `CallTargetThunk` can just "semantically" represent the tail of the method except it gets its arguments from somewhere else.

janvorli replied:
@jakobbotsch I'm not sure how we could make it work without the calli. When making a tail call with helper, we first call the StoreArguments which stores all the arguments including the "this" pointer and the generic pointer argument. The CallTarget then gets the buffer with the stored arguments, extracts them and passes them to the target function. If we used call instead of calli here, how could we pass in the extracted generic argument?

jakobbotsch replied:
I can see why we may have problems since we somehow will have to pass the generic context through to the call thunk, and we may not have enough stack space to do this (only a problem on x86). I wonder if this could be special cased in the x86 helper instead, by considering it to have __fastcall calling convention. However it would have to push it on the stack before calling the call thunk.

jakobbotsch replied:
On other platforms/ABIs this is what I would imagine, taking @jkotas's example from above:
```csharp
static void foo<T>()
{
    ...
    tailcall.    bar<List<T>>(1,2,3,4,5,6,7,8);
}

static void bar<T>(int a1, int a2, int a3, int a4, int a5, int a6, int a7, int a8)
{
   ....
}
```

We turn this into:
```
static void foo<T>()
{
    ...
    call StoreBarArguments(1, 2, 3, 4, 5, 6, 7, 8);
    tailcall.    CallBar<T>();
}

static void CallBar<T>()
{
    bar<List<T>>(args...);
}
```

and JIT treats the tailcall to `CallBar` exactly as normal except it uses the tail call helper as its target. On other ABIs the tailcall to CallBar here is always possible.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
````diff
@@ -0,0 +1,375 @@
+* Callee saved registers and locals of D
+```
+### Returned to CallTarget thunk for D with ChainCall flag in the TailCallHelperStackEntry **clear**
+The thunk checks the ChainCall flag and since it is clear, it recognizes we are now returning from the call chain and so it returns the result of the D. 
````
Is this check of ChainCall flag, etc. missing in the thunk IL example below?

janvorli replied:
Ah, yes, I have forgotten to add it in there.

janvorli commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,375 @@
+    ldloc.0
+    ldobj object // T t
+
+    calli Point (object, object, int32, object)
```
Ah, the address loading is missing in this example. I have added these examples as a last minute idea and now I can see I should have spent more time on them. 
The calli is there so that the call can be blind and we can pass this and generic context explicitly (I may be wrong, but I think the regular call doesn't allow this)

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
````diff
@@ -0,0 +1,375 @@
+For tail calling the Test function, the IL helpers could look e.g. as follows:
+
+```IL
+.method public static Point StoreArgumentsTest(object thisObj, object genericContext, int x, object t) cil managed
````
>object thisObj, [](start = 47, length = 15)

Since Point is a struct, thisObj is actually a byref instead of an obj-ref.

janvorli replied:
@sivarv thisObj is Foo in this case

sivarv commented on Documentation/design-docs/tailcalls-with-helpers.md:
````diff
@@ -0,0 +1,375 @@
+```
+
+```IL
+.method public static Point CallTargetTest() cil managed
````
When would CallTarget IL thunk gets generated?  On demand when TailCallHelper needs to call CallTarget?

Similarly when is StoreArgument IL thunk gets generated?  When JIT asks for its method hnd or entry-point addr?

janvorli replied:
They need to be generated at the same time - the StoreArgument stores the address of the CallTarget into the buffer (forgotten that in the example). I was thinking that it would be at the time jit asks for its method hnd.

sivarv commented:
Looks good to me.

danmosemsft commented:
@janvorli sounds like this is good to merge?

janvorli commented:
@danmosemsft no, it is still in progress. I need to update the document based on the last feedback and there is still one detail I'll need to verify that I have discovered while making changes based on the feedback.

janvorli commented:
@sivarv, @jkotas I have updated the document based on your feedback. I have corrected the IL helper examples to contain all the pieces that were missing and added the tail call target address to the arguments buffer and as a parameter to the StoreArguments. I have also changed the return type of the StoreArguments to be void to get rid of the need to create and return the dummy value.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,439 @@
+    call native int GetCurrentSp()
+    stfld native int TailCallHelperStackEntry::sp
+    ldloca.s 1
+    ldfld native int t_helperStackTOS
```
I do not think this should be a managed field. It needs to have a special handling during in the GC, and making it a managed field will make that hard. The push/pop should be calls to runtime helpers.

janvorli replied:
@jkotas You mean the t_helperStackTOS? Yes, I agree. I have put it this way into the example thinking it would be more obvious how it works. I'll change the example to match the reality.

jkotas commented:
We will also want to make sure that debugger stepping works for this well - TailCallStubManager will need to be updated accordingly.

janvorli commented:
Good point. I'll add it to the list work that needs to be done.

janvorli commented:
@jkotas I've augmented the last commit based on your last two comments.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,435 @@
+{
+    .maxstack 4
+
+    call native int FetchArgumentBuffer()
```
Shouldn't this be called AllocateArgumentBuffer and take the Argument GC descriptor?

janvorli replied:
Yes, it should. This is just another remainder from my experimental hand written tests where I had a preallocated fixed size buffer for the sake of simplicity.

jkotas commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,435 @@
+    sizeof object
+    add
+
+    dup
```
Nit: These chained dups+adds are not be very friendly for the JIT. Each of them is likely creating a new local. Having one explicit local - kind of like what's in the other stub - that you keep incrementing would be more JIT friendly.

janvorli replied:
Ok, good to know. I guess it would be even better just to use absolute offsets instead of incremental ones.

jakobbotsch commented:
I have made a simple functional example/prototype of this approach. It computes whether 10 million is even using mutual recursion such that `IsOdd` cannot tailcall `IsEven` (on Windows). The code is here:
https://gist.github.com/jakobbotsch/a49a651a5ef77f1d5bceb0b702f9de5a

The baseline cheats a little since it does not need to copy and pass an extra GUID. Also, I am missing logic to push/pop an argument descriptor/buffer and just use a single thread local instead, so the real thing would be somewhat more expensive.

That said, the results on my PC are encouraging for this approach:
```
IsOddFast: False in 27 ms
IsOddOldHelper: False in 1298 ms
IsOddNewHelper: False in 119 ms
```

jakobbotsch commented on Documentation/design-docs/tailcalls-with-helpers.md:
```diff
@@ -0,0 +1,460 @@
+    stfld bool TailCallHelperStackEntry::chainCall
+    ldloca.s 1
+    call native int GetCurrentSp()
+    stfld native int TailCallHelperStackEntry::sp
```
How do we envision this to be used by the helper and the JIT? Will `GetCurrentSp` be an intrinsic? In my experiment I have implemented the helper with a simple check that `rsp` is equal to this value to figure out whether the current call frame is this one. However `rsp` will be after the `calli` below which is not always trivial to predict (although due to `FEATURE_FIXED_OUT_ARGS` in the JIT it should be for many platforms).

I wonder if it would be easier to use the return address although I don't see exactly how we could do that either.

janvorli replied:
It could be an intrinsic or a JIT helper. It seems the perf difference would be negligible due to the other stuff we need to do to perform the tail call. 
The SP we get in the TailCallHelper should actually not be influenced by the calli parameters even on x86, since on that platform, the callee cleans the stack, so when the TailCallHelper is tail-called by a function that was tail-called too, the arguments should be already removed from the stack and so the SP should match the captured one.

Btw, I have actually planned to use the return address instead in one of the initial iterations of the design. My original plan was that the tail call helper stack entry would store the range of IP addresses of the CallHelper so that I can compare the return address to that range. My guess is that the reason for the switch was that I didn't know if JIT, even with an addition of some intrinsic, can reasonably figure that out. And the SP capturing  / comparison was simpler.

jakobbotsch replied:
> The SP we get in the TailCallHelper should actually not be influenced by the calli parameters even on x86, since on that platform, the callee cleans the stack, so when the TailCallHelper is tail-called by a function that was tail-called too, the arguments should be already removed from the stack and so the SP should match the captured one.

Good point, I have not realized that. 

> Btw, I have actually planned to use the return address instead in one of the initial iterations of the design. My original plan was that the tail call helper stack entry would store the range of IP addresses of the CallHelper so that I can compare the return address to that range. My guess is that the reason for the switch was that I didn't know if JIT, even with an addition of some intrinsic, can reasonably figure that out. And the SP capturing / comparison was simpler.

I still wonder if we can perhaps solve both the problems with generic context and return address using a single intrinsic. Something like
```
ldtoken TargetFunction
ldloc tailCallStackEntry
ldflda TailCallStackEntry::retAddr
ld Args::genericContext
ld Args::retBuffer
ld Args::arg1
ld Args::arg2
...
ld Args::argN
call System.Runtime.JitTailCallDispatchDirect
```
and the JIT would store the return address into `retAddr` and do the necessary setup for the generic context and retbuffer. Initial prototype would probably only have one indirect dispatch intrinsic, but later the direct version (shown above) could be added too allowing for inlining if that was deemed worthwhile.

By the way, I plan to work on implementing (a prototype of) the design in this PR for the rest of my internship.

jakobbotsch commented:
I have experimented with a couple of variants of the design from here. I have updated my gist:
https://gist.github.com/jakobbotsch/a49a651a5ef77f1d5bceb0b702f9de5a

The results are as follows:
```
Fast tailcalls, 300 runs: min: 14.71 ms, max: 64.57, avg: 20.50
Old helper, 5 runs: min: 1311.87 ms, max: 1582.45, avg: 1433.36
With new ASM helper, 100 runs: min: 137.56 ms, max: 391.81, avg: 181.88
No ASM helper, 100 runs: min: 100.73 ms, max: 183.60, avg: 125.77
Looping, 100 runs: min: 141.41 ms, max: 229.46, avg: 155.67
```
They are all using `[ThreadStatic]`, which in the real implementation would probably be faster native TLS instead.

To gauge the impact of TLS, here are the results when the "thread static buffer" is not actually marked as `[ThreadStatic]`:
```
Fast tailcalls, 300 runs: min: 15.08 ms, max: 51.70, avg: 19.83
Old helper, 5 runs: min: 1274.29 ms, max: 1516.84, avg: 1403.04
With new ASM helper, 100 runs: min: 78.92 ms, max: 136.55, avg: 85.48
No ASM helper, 100 runs: min: 55.86 ms, max: 97.20, avg: 63.02
Looping, 100 runs: min: 52.82 ms, max: 97.05, avg: 60.51
```

As can be seen there is significant overhead from the TLS, but it is still much better than the previous helper.

"With new ASM helper" is the design as described in this document. "No ASM helper" is a different design that does not require the new ASM helper. Instead of compiling
```csharp
return tail Foo(a, b, c, d)
```
into 
```csharp
StoreArgsFoo(a, b, c, d);
return tail TailCallHelper()
```

it is compiled into
```csharp
StoreArgsFoo(a, b, c, d);
return tail CallFoo();
```

and `CallFoo` is modified to check if there is a "CallTarget" stub in the previous frame, so that it looks something like the following:

```csharp
static T CallFoo()
{
    var tos = GetTopOfTailCallStack();

    if (PrevFrameIsCallTargetStub(tos))
    {
        tos->NextCall = &CallFoo;
        return false;
    }

    TailCallHelperStackEntry newEntry;
    newEntry.prev = prevEntry;
    newEntry.StackPointer = GetStackPointer();
    newEntry.NextCall = IntPtr.Zero;
    PushTailCallFrame(&newEntry);

    T result;
    try
    {
        result = Foo(...args through TLS...);
    }
    finally
    {
        PopTailCallFrame();
    }

    if (newEntry.NextCall != IntPtr.Zero)
    {
        return tail newEntry.NextCall();
    }

    return result;
}

```

The benefit of this is that we do not need a per-platform ASM helper, although we do need to implement `PrevFrameIsCallTargetStub` somehow which will probably involve some JIT intrinsics. Also the runtime will need to have minimal knowledge of the stack.

"Looping" is yet another variant which tries to avoid much of the pushing/popping of frames. We transform
```csharp
return tail Foo(a, b, c, d)
```
into 
```csharp
StoreArgsFoo(a, b, c, d);
return tail DispatchTailCalls<T>();
```

Where `DispatchTailCalls` is a function implemented in CoreLib. It checks if the previous frame is another dispatcher, and in that case returns to it. It will look like the following (from my gist):
```csharp
static bool DispatchTailCalls(int a)
{
    ref ThreadStaticBuffer buf = ref GetBuffer();
    var prevEntry = (TailCallHelperStackEntry*)buf.TCFrames;

    if ((IntPtr)(&a) == prevEntry->StackPointer)
        return false; // Let upper dispatcher deal with it

    TailCallHelperStackEntry frame;
    frame.prev = prevEntry;

    buf.TCFrames = (IntPtr)(&frame);
    bool result;
    try
    {
        do
        {
            IntPtr saved = buf.Args.CallThunk;
            buf.Args.CallThunk = IntPtr.Zero;
            IL.Push(saved);
            IL.Emit.Calli(new StandAloneMethodSig(CallingConventions.Standard, typeof(bool)));
            IL.Pop(out result);
        } while (buf.Args.CallThunk != IntPtr.Zero);
    }
    finally
    {
        buf.TCFrames = (IntPtr)frame.prev;
    }

    return result;
}

```

This makes `CallTarget` stubs somewhat simpler, but it is somewhat harder to implement due to handling return values in `DispatchTailCalls`. I will continue with the second last variant (which does away with the helper but retains most of the rest of the design).
